src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -82,51 +79,29 @@
                                           (filter dir-create?)
                                           (map #(.toPath (:path %)))))))
 
-(defn reset-key-and-unwatch-deleted!
-  [watch-key]
-  ;; reset key and remove from set if directory no longer accessible
-  (let [valid? (.reset watch-key)]
-    (when (not valid?)
-      ;; TODO: not sure if we need to do anything else here now that we're not retaining a map of
-      ;; these keys.  The Oracle example code makes it seem like maybe you do, but I can't tell what.
-      (log/info (trs "Removing watched directory {0} because it was deleted" (.watchable watch-key))))))
-
-;; This function (and other helper functions in this namespace)
-;; are heavily influenced by WatchDir.java from Java's official documentation:
-;; https://docs.oracle.com/javase/tutorial/essential/io/notification.html
-(schema/defn handle-watch-events!
-  [watchers :- [(schema/protocol Watcher)]
+(schema/defn watch!
+  "Returns a future representing processing events for the passed in watcher.
+  The future will swallow errors, loop producing side-effects until stopped,
+  and is expected to coordinate with other functions via the stopped? atom."
+  [watcher :- (schema/protocol Watcher)
+   stopped? :- Atom
    shutdown-on-error :- IFn]
-
-  ;; If we were going to add any additional "debouncing" here I think it
-  ;; would look something like this:
-  ;;
-  ;(let [mas? (atom true)
-  ;      watch-keys (atom [])]
-  ;  (while @mas?
-  ;    (if-let [watch-key (.poll watch-service)]
-  ;      (swap! watch-keys conj watch-key)
-  ;      (reset! mas? false))))
-  ;;
-  ;; (yes, I know that code is ugly as sin, and actually, as Chris points out,
-  ;; perhaps what we'd want to do is something like write a small implementation
-  ;; of Java's Iterator interface that wrapped the call to .poll,
-  ;; and then you could use Clojure's iterator-seq.)
-  ;;
-  ;; ... The idea being that we'd collect all of the available watch keys
-  ;; first, and then process them as a batch.  I think this could result
-  ;; in fewer callback invocations for nested watch keys (i.e. files
-  ;; nested under the root watched path) but the extent to which that is
-  ;; true is likely platform-dependent.
-  ;;
-  ;; https://tickets.puppetlabs.com/browse/PE-15621
-
-  (doseq [watcher watchers]
-    (when-let [watch-key (.poll (:watch-service watcher))]
-      (let [orig-events (.pollEvents watch-key)
+  (future
+    (while (not @stopped?)
+      ;; `.take` here will block until there has been an event for this key
+      ;;
+      ;; If `.close` is called on the `WatchService` while blocking on
+      ;; `.take` a `ClosedWatchServiceException` will be raised.
+      ;; The `WatchKey` will then be invalidated, so `.reset` will return
+      ;; false, but enqueued events will still be available via `.pollEvents`
+      ;; and `.watchable` should continue to function as desired.
```

Comment:
Your comment makes sense, but I think it would be better to not execute the rest of this loop if this exception is caught.

If this exception is caught, that means that the service's `stop` function executed and the application is shutting down.  It's possible that the other services which registered callbacks have already shutdown themselves, in which case, invoking their event handler callbacks might not even be a safe thing to do.

The code is certainly a little cleaner like this, but I don't think it's the correct behavior.  It might be sufficient to simply move the rest of the body of this function inside `try`.